### ISSUE TITLE
Mappable RwLock guards

### ISSUE THREAD

**Issue by kvark** [Amanieu/parking_lot#83]
This is a logical continuation of  #10. It looks like the `MappableMutexGuard` is implemented, but nothing is done for `RwLock`. We have a strong case for it: a map with data that needs to be either read or created and then read. Ideally, it would have the following implementation (roughly):
```rust
fn get_data(&self, key: Key) -> ReadGuard<Data> {
  // lock the map, but only deal with a particular entry
  let guard = self.lock.upgradable_read().map(|map| map.entry(key));
  // fast path for keys already filled up
  if let Entry::Occupied(e) = *guard {
    return *e;
  }
  // slow path to add a new value
  let data = create_heavy_state(key);
  let gw = guard.upgrade();
  gw.or_insert(data);
  // and return it
  gw.downgrade().or_insert_with(|| ureachable!())
}
```

**Comment by Amanieu** [comment]
If you look at the [docs](https://docs.rs/parking_lot), `map` is implemented for `RwLockReadGuard` and `RwLockWriteGuard`.

However it is not implemented for `RwLockUpgradableReadGuard` because there are soundness issues with upgrading such a map. This is because the closure passed to `map` can return any arbitrary `&T`, and it is unsound to assume that this can safely be upgraded to `&mut T` for all possible reference types.

**Comment by Amanieu** [comment]
Looking at your example, you might want to defer calling `map` until after you have downgraded the lock to a normal `RwLockReadGuard`.

**Comment by kvark** [comment]
@Amanieu indeed the non-upgradable ones have `map`, I missed this initially. I've spent quite a bit of time trying to implement anything close to the pseudo-code written here, and I wasn't successful. I figured we need to implement our own map storage type for this.

**Comment by kvark** [comment]
@Amanieu I think you understand my use case. Getting a read lock followed by a possible write lock has the following downsides for me:
  1. need to re-acquire the lock, although I think this is the least of concerns
  2. accessing the map by key twice, even though this only happens when we need to add an entry, so the cost is amortized and we don't optimize for this case anyway
  3. impossible to abstract away this in a nice self-contained API, e.g.
```rust
fn get_or_create_with<F: FnOnce() -> V>(lock: RwLock<Self>, key: Key, create_fn: F) -> &V {...}
```

In other words, I'd like this function to borrow the lock immutably and return something that can be de-referenced into the value. If this function tries to mess with locking internally in any way, there is no ability to return such a value wrapper any more.

**Comment by Amanieu** [comment]
@kvark Here's a quick sketch of how this would look like:

```rust
fn get_or_create_with<F: FnOnce() -> V>(lock: &RwLock<Self>, key: Key, create_fn: F) -> MappedRwLockReadGuard<V> {
    {
        let guard = RwLockReadGuard::map(lock.read(), |map| map.get(key));
        if guard.is_some() {
            return RwLockReadGuard::map(guard, |opt| opt.unwrap());
        }
    }

    RwLockWriteGuard::map(lock.write(), |map| map.entry(key).or_insert_with(create_fn)).downgrade()
}
```

**Comment by kvark** [comment]
@Amanieu 
This code is very interesting, thank you! I'm not yet able to make it work though.
In particular, a problem I'm seeing here is that functions passed to `::map` calls require to return references. This is not the case for either `HashMap::get` (which returns `Option<&V>`, which is not a reference) or `HashMap::entry` (which is `Entry`). Thus, the lock mapping API appears to be fundamentally incompatible with maps, among other things.

**Comment by kvark** [comment]
Also cc-ing @Gankro on the last comment here ^

**Comment by Amanieu** [comment]
@kvark Oh, good point, I missed that. It seems that there is no way to avoid a double lookup with the current API then:

```rust
let guard = lock.read();
if guard.contains_key(key) {
    return RwLockReadGuard::map(guard, |map| map.get(key).unwrap());
}
```

I think we could add a `try_map` API for this use case, if there is sufficient motivation.

**Comment by kvark** [comment]
@Amanieu heads up, we happen to need `try_map` now for `webgpu-remote` component of https://github.com/gfx-rs/wgpu

**Comment by Amanieu** [comment]
It it now available in `lock_api` 0.1.4, which `parking_lot` depends on.
